File: backend/app/routers/games.py
import uuid
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.database import get_db
from app.games.registry import list_games, get_game
from app.models.conversation import Conversation
from app.models.user import User
from app.schemas.chat import GameResponse, MessageResponse
from app.services.auth_service import get_current_user

router = APIRouter(prefix="/api/games", tags=["games"])


@router.get("", response_model=list[GameResponse])
async def get_games(language: str = Query(default="en")):
    """List all available games, with descriptions in the requested language."""
    games = list_games()
    return [
        GameResponse(
            id=g.id,
            name=g.name,
            description=g.get_description(language),
            initial_mood=g.initial_mood,
            initial_message=g.get_initial_message(language),
        )
        for g in games
    ]


class PasswordCheckRequest(BaseModel):
    game_id: str
    password: str
    conversation_id: uuid.UUID | None = None


class PasswordCheckResponse(BaseModel):
    correct: bool


@router.post("/check-password", response_model=PasswordCheckResponse)
async def check_password(
    request: PasswordCheckRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Check if the submitted password matches the game's secret password."""
    game = get_game(request.game_id)
    if not game or not game.password:
        return PasswordCheckResponse(correct=False)

    correct = request.password.strip().upper() == game.password.strip().upper()

    # If correct and we have a conversation_id, mark it as solved
    if correct and request.conversation_id:
        result = await db.execute(
            select(Conversation).where(
                Conversation.id == request.conversation_id,
                Conversation.user_id == current_user.id,
            )
        )
        conversation = result.scalar_one_or_none()
        if conversation and not conversation.solved_at:
            conversation.solved_at = datetime.now(timezone.utc)
            await db.flush()

    return PasswordCheckResponse(correct=correct)


class SolvedConversationResponse(BaseModel):
    id: uuid.UUID
    title: str
    username: str
    solved_at: datetime
    created_at: datetime
    messages: list[MessageResponse] = []

    model_config = {"from_attributes": True}


@router.get("/{game_id}/hall-of-fame", response_model=list[SolvedConversationResponse])
async def get_hall_of_fame(
    game_id: str,
    db: AsyncSession = Depends(get_db),
):
    """List all solved conversations for a game (Hall of Fame), visible to everyone."""
    result = await db.execute(
        select(Conversation)
        .where(
            Conversation.game_id == game_id,
            Conversation.solved_at.isnot(None),
        )
        .options(selectinload(Conversation.messages), selectinload(Conversation.user))
        .order_by(Conversation.solved_at.asc())
    )
    conversations = result.scalars().all()

    return [
        SolvedConversationResponse(
            id=c.id,
            title=c.title,
            username=c.user.username,
            solved_at=c.solved_at,
            created_at=c.created_at,
            messages=[
                MessageResponse(
                    id=m.id,
                    role=m.role,
                    content=m.content,
                    model=m.model,
                    prompt_tokens=m.prompt_tokens,
                    completion_tokens=m.completion_tokens,
                    created_at=m.created_at,
                )
                for m in c.messages
            ],
        )
        for c in conversations
    ]

File: frontend/src/services/games.ts
import type {Game} from '../types'

/**
 * Fetch the list of available games from the backend.
 * Optionally pass a language code to get translated descriptions.
 */
export async function fetchGames(language?: string): Promise<Game[]> {
    const token = localStorage.getItem('access_token')

    const params = new URLSearchParams()
    if (language) {
        params.set('language', language)
    }
    const qs = params.toString()
    const url = `/api/games${qs ? `?${qs}` : ''}`

    const response = await fetch(url, {
        headers: {
            Authorization: `Bearer ${token}`,
        },
    })

    if (!response.ok) {
        if (response.status === 401) {
            window.location.href = '/login'
            return []
        }
        throw new Error(`Failed to fetch games: ${response.statusText}`)
    }

    return response.json()
}

/**
 * Check if a password guess is correct for a password-lock game.
 */
export async function checkPassword(gameId: string, password: string, conversationId?: string | null): Promise<boolean> {
    const token = localStorage.getItem('access_token')

    const body: Record<string, string> = { game_id: gameId, password }
    if (conversationId) {
        body.conversation_id = conversationId
    }

    const response = await fetch('/api/games/check-password', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            Authorization: `Bearer ${token}`,
        },
        body: JSON.stringify(body),
    })

    if (!response.ok) {
        throw new Error(`Failed to check password: ${response.statusText}`)
    }

    const data = await response.json()
    return data.correct
}

export interface SolvedConversation {
    id: string
    title: string
    username: string
    solved_at: string
    created_at: string
    messages: Array<{
        id: string
        role: string
        content: string
        model?: string
        created_at: string
    }>
}

/**
 * Fetch the Hall of Fame — all solved conversations for a game.
 */
export async function fetchHallOfFame(gameId: string): Promise<SolvedConversation[]> {
    const token = localStorage.getItem('access_token')

    const response = await fetch(`/api/games/${gameId}/hall-of-fame`, {
        headers: {
            Authorization: `Bearer ${token}`,
        },
    })

    if (!response.ok) {
        throw new Error(`Failed to fetch hall of fame: ${response.statusText}`)
    }

    return response.json()
}
